File: packages/core/src/parser/__testdata__/sample.pl
#!/usr/bin/perl
package Synapse::Agent;

use strict;
use warnings;
use Carp;
use JSON::XS;
use LWP::UserAgent;

use constant MAX_RETRIES => 3;
use constant DEFAULT_MODEL => 'claude-opus-4-6';

our $VERSION = '1.0.0';
my $agent_count = 0;

sub new {
    my ($class, %args) = @_;
    my $self = bless {
        name        => $args{name} || croak("name required"),
        model       => $args{model} || DEFAULT_MODEL,
        max_tokens  => $args{max_tokens} || 4096,
        status      => 'idle',
        _tools      => ['search', 'read', 'write'],
    }, $class;
    $agent_count++;
    return $self;
}

sub process {
    my ($self, $message) = @_;
    croak("Empty message") unless defined $message && length $message;
    $self->{status} = 'active';
    my $result = $self->_call_model($message);
    $self->{status} = 'idle';
    return $result;
}

sub get_tools {
    my ($self) = @_;
    return @{$self->{_tools}};
}

sub status {
    my ($self) = @_;
    return $self->{status};
}

sub _call_model {
    my ($self, $message) = @_;
    # TODO: implement actual API call
    return "Response to: $message";
}

sub DESTROY {
    my ($self) = @_;
    $agent_count--;
}

sub _validate {
    my ($self, $input) = @_;
    return defined $input && length($input) > 0;
}

1;

# FIXME: add proper error handling

File: packages/core/src/parser/__testdata__/sample.dart
import 'dart:async';
import 'dart:convert';
import 'package:meta/meta.dart';

const int maxRetries = 3;
const String defaultModel = 'claude-opus-4-6';

enum Status { active, idle, stopped, error }

class AgentConfig {
  final String model;
  final int maxTokens;
  final double temperature;

  const AgentConfig({
    this.model = defaultModel,
    this.maxTokens = 4096,
    this.temperature = 0.7,
  });

  factory AgentConfig.fromJson(Map<String, dynamic> json) {
    return AgentConfig(
      model: json['model'] as String? ?? defaultModel,
      maxTokens: json['max_tokens'] as int? ?? 4096,
    );
  }
}

abstract class Agent {
  String get name;
  Status get status;
  Future<String> process(String message);
  List<String> getTools();
}

mixin HasLogging {
  void log(String message) => print('[Agent] $message');
}

abstract class BaseAgent extends Agent with HasLogging {
  @override
  final String name;
  AgentConfig config;
  Status _status = Status.idle;

  BaseAgent(this.name, {AgentConfig? config})
      : config = config ?? const AgentConfig();

  @override
  Status get status => _status;

  @protected
  bool validate(String input) => input.trim().isNotEmpty;
}

class SynapseAgent extends BaseAgent {
  SynapseAgent(super.name, {super.config});

  @override
  Future<String> process(String message) async {
    if (!validate(message)) throw ArgumentError('Empty message');
    _status = Status.active;
    final result = await _callModel(message);
    _status = Status.idle;
    return result;
  }

  @override
  List<String> getTools() => ['search', 'read', 'write'];

  Future<String> _callModel(String message) async {
    // TODO: implement actual API call
    return 'Response to: $message';
  }

  factory SynapseAgent.create(String name) => SynapseAgent(name);
}

extension AgentExtensions on Agent {
  bool get isActive => status == Status.active;
  bool get isIdle => status == Status.idle;
}

// FIXME: add proper error handling
